src/test/java/org/stratumproject/fabric/tna/behaviour/upf/MockP4RuntimeClient.java

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022-present Open Networking Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.stratumproject.fabric.tna.behaviour.upf;
+
+import com.google.common.collect.Maps;
+import org.onosproject.net.DeviceId;
+import org.onosproject.net.pi.model.PiPipeconf;
+import org.onosproject.net.pi.runtime.PiCounterCell;
+import org.onosproject.net.pi.runtime.PiCounterCellData;
+import org.onosproject.net.pi.runtime.PiCounterCellId;
+import org.onosproject.net.pi.runtime.PiPacketOperation;
+import org.onosproject.p4runtime.api.P4RuntimeClient;
+
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.LongStream;
+
+import static org.stratumproject.fabric.tna.behaviour.P4InfoConstants.FABRIC_EGRESS_UPF_TERMINATIONS_COUNTER;
+import static org.stratumproject.fabric.tna.behaviour.P4InfoConstants.FABRIC_INGRESS_UPF_TERMINATIONS_COUNTER;
+
+/**
+ * Used to mock P4Runtime client used only for read/write requests for counters.
+ */
+public class MockP4RuntimeClient implements P4RuntimeClient {
+
+    private final DeviceId deviceId;
+    private final Map<Long, PiCounterCell> igCounters;
+    private final Map<Long, PiCounterCell> egCounters;
+
+    /**
+     * Used to mock P4Runtime client.
+     *
+     * @param deviceId    The ID of the device
+     * @param counterSize The size of the counter array
+     */
+    public MockP4RuntimeClient(DeviceId deviceId, int counterSize) {
+        this.deviceId = deviceId;
+        igCounters = Maps.newHashMap();
+        egCounters = Maps.newHashMap();
+        LongStream.range(0, counterSize).forEach(i -> {
```

Comment:
I am confused - what is it doing this code 😶‍🌫️  ?

Reply:
ops